pkgs/development/python-modules/openai-triton/default.nix

```diff
@@ -111,6 +112,11 @@ buildPythonPackage rec {
     # Use our linker flags
     substituteInPlace python/triton/common/build.py \
       --replace '${oldStr}' '${newStr}'
+    # triton/common/build.py will be called both on build, and sometimes in runtime. This is cursed.
+    substituteInPlace python/triton/common/build.py \
+      --replace 'os.getenv("TRITON_LIBCUDA_PATH")' '"${cudaPackages.cuda_cudart}/lib"'
```

Comment:
The environment variable says "libcuda" (the userspace driver), not "libcudart"?

> This is cursed.

Yes and this is intentional: isn't triton literally a tool for compiling kernels on the fly from some subset of python?

Reply:
Also, wasn't there an attempt to make CUDA stuff optional in triton? In that case we don't want to refer to backendStdenv but to the conditional stdenv (otherwise the cpu-only version pulls two different GCCs into the closure)

Reply:
Yes, sorry for confusion, I've built in on some other base, and those changes are not working as-is, I'll redo this PR properly for upstreaming to nixpkgs.

Reply:
> The environment variable says "libcuda" (the userspace driver), not "libcudart"? Is this a confusion on upstream's side?

It wants libcuda.so.1, I'm not sure where I need to look for it?

> Yes and this is intentional: isn't triton literally a tool for compiling kernels on the fly from some subset of python?

The cursed part is that build and runtime step are closely intermixed, but the build step doesn't have a way to provide some values for runtime.
It provides ptxas as third_party binary, but libcuda and etc are expecting that it is running with the same binaries as it built with, while this is not necessary true, as I'm sure some things there are not ABI-compatible. I see in 3.0.0 version the build process is making much more sense.

> Also, wasn't there an attempt to make CUDA stuff optional in triton? In that case we don't want to refer to backendStdenv but to the conditional stdenv (otherwise the cpu-only version pulls two different GCCs into the closure)

The CC variable read override is only enabled on cudaSupport, otherwise it doesn't try to call CC at runtime, and the build-time CC is enough (At least, I haven't experienced that with vLLM)

Reply:
> It wants libcuda.so.1, I'm not sure where I need to look for it?

Libcuda depends on the (nvidia) kernel (module) that runs on the user's machine, so we don't link it through the nix store, we link it through `/run/opengl-driver/lib`: `${addDriverRunpath.driverLink}/lib`.

More specifically, we use the fake driver `${getLib cudaPackages.cuda_cudart}/lib/stubs` at build/link time, and `${addDriverRunpath.driverLink}/lib` at runtime. It's also important that at runtime we first try to `dlopen("libcuda.so", ...)` first, and only then `dlopen("/run/opengl-driver/lib/libcuda.so", ...)` because we want things to also work on FHS distributions and respect the optional `LD_LIBRARY_PATH`

> The cursed part is that build and runtime step are closely intermixed, but the build step doesn't have a way to provide some values for runtime.

Oh right, we should probably try to explicitly track the references retained at runtime.

> The CC variable read override is only enabled on cudaSupport, otherwise it doesn't try to call CC at runtime, and the build-time CC is enough (At least, I haven't experienced that with vLLM)

Do they not use `triton.common.build` at runtime for their jit/aot?

Reply:
> More specifically, we use the fake driver ${getLib cudaPackages.cuda_cudart}/lib/stubs at build/link time, and ${addDriverRunpath.driverLink}/lib at runtime. It's also important that at runtime we first try to dlopen("libcuda.so", ...) first, and only then dlopen("/run/opengl-driver/lib/libcuda.so", ...) because we want things to also work on FHS distributions and respect the optional LD_LIBRARY_PATH

I don't think we can achieve that with this code? It runs both at compile time, and at runtime...
Except by patching it in preInstall?..

> Do they not use triton.common.build at runtime for their jit/aot?

Not in vLLM, I'm not sure about other projects using triton directly.

Reply:
Well we should open an issue asking for a more fine-grained support. Note they do not use the variable on `master` any more, but use `whereis` which is also platform-specific: https://github.com/feihugis/triton/blob/a9d1935e795cf28aa3c3be8ac5c14723e6805de5/python/triton/compiler.py#L1354-L1357